### PR TITLE
store-gateway: don't wait on lazy loading if queries are canceled

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
#### Goal

Allows requests to abort waiting for lazy loading to complete if the request is cancelled. However, we don't want the `LazyBinaryReader` to abort loading the binary reader.

#### Challenge

The main challenge with the current implementation was the usage of mutexes. Doing this with mutexes is impractical because they block until the mutex is acquired and there's no way to abort waiting (without doing busy loops with `TryLock()`).

#### Implementation

Because of this challenge I chose to reimplement the synchronisation in the `LazyBinaryReader` with a control loop rather than a mutex (see `LazyBinaryReader.controlLoop`).
There is a single goroutine which ordered the reads and writes (load/unload) of the underlying binary reader.
Interacting with the state happens via sending a callback to a channel (a request).
The control goroutine receives callbacks/requests and manages the state of the underlying reader. When it's done processing the request it sends a response to the request.
Sending the request and receiving the response can both be aborted by the client upon context cancellation.

The only exception is storing the `lastUsedAt` timestamp for the reader in an atomic. The reason for this is that access to that should be fast rather than accurate. The timestamp is accessed when building a snapshot for all the used blocks. That happens under [`lazyReadersMx`](https://github.com/grafana/mimir/blob/40ee83bd595aa2316d40e5e24047a9d6cb257955/pkg/storegateway/indexheader/reader_pool.go#L263-L273), which is in turn necessary for adding new blocks. Having `lastUsedAt` be managed by the control loop means that we might block adding new blocks because one block is slow to lazy load, which is hard to debug and generally something we want to avoid.

#### Side effect 

I also noticed another inefficiency that this PR solves unintentionally. Before this PR the [lazy loading gate is acquired
before acquiring the write lock for loading the reader](https://github.com/grafana/mimir/blob/04e6b7fc0f305f344b0e487691218d898633410a/pkg/storegateway/indexheader/lazy_binary_reader.go#L274-L281). This means that if there are multiple clients (queries)
trying to load the same block, they can all consume the concurrency quota, but only one of them will actually do the loading.
The rest of the clients will occupy a slot in the concurrency gate without doing anything.

This PR resolves this by making sure that only the control goroutine tries to load an index header.

#### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir/issues/8257

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [inline pkg/storegateway/indexheader/lazy_binary_reader.go:172]: this came up in https://github.com/grafana/mimir/pull/8281#discussion_r1638174462 as well. Either @narqo or I will take this up to clean up the leaking goroutines soon (next week?)

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [inline pkg/storegateway/indexheader/lazy_binary_reader.go:84]: Nit: I think that `C` (`ch`, `chan`, etc.) suffixes on channel variable/fields are redundant with the type system we have.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [inline pkg/storegateway/indexheader/lazy_binary_reader.go:278]: It seems that every time we try to `getReader`, we'll load one if there isn't one yet. 

So why is this logic here, instead of having just a single channel to get the reader, and let the `controlLoop` take care of loading one if needed?

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [inline pkg/storegateway/indexheader/lazy_binary_reader.go:387]: It seems that once the error is stored, nothing cleans it, is that right? A transient error (disk is full), would prevent this block from being loaded ever again.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:172]: I feel that "a proper service lifecycler" would be an overkill here. Can we just pass a param in the unload request to return from the `controlLoop`, and pass it in the `Close()` call?

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:387]: i had the same doubt. I erred on the side of retaining existing behaviour. I don't see why we'd want to retain the error to be honest.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:278]: oh, good point, this makes it much cleaner. I will refactor

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:377]: the only serial part here is obtaining a reader. You can obtain multiple readers and use them parallel.

Granted, it will be slower to obtain 100 readers because you need to do 200 channel send+receive serially compared to the 100 concurrent RLock+RUnlocks before. But my gut feeling is that it is negligible in the scheme of the i/o-intensive work that the store-gateway does.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:387]: I will take another look and just enable retries if there are no negative consequences.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:307]: that's how it was initially, but the rest of the methods take a struct, so I went with consistency. Also nesting a few channels makes it a bit hard to read (`chan chan loadedReader` looked less familiar to me; `chan readerRequest` looked a bit easier)

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [inline pkg/storegateway/indexheader/lazy_binary_reader.go:194]: Opinionated ignorable nit: I would wrap the `Reader` returned by `getOrLoadReader` to:

```go
type lazyReader struct {
    Reader
    wg *sync.WaitGroup
}

func (lr lazyReader) release() {
    lr.wg.Done()
}
```

So here we could just do:
```suggestion
	reader, err := r.getOrLoadReader(ctx)
	if err != nil {
		return 0, err
	}
	defer reader.release()
```

Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:194]: i totally agree. I just didn't want to extend the changeset further; i can do this after we merge

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:377]: i don't think we'll be A-B testing this, so I can try to run a few benchmarks before merging

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (seizethedave) [inline pkg/storegateway/indexheader/lazy_binary_reader.go:157]: A lot of complexity and fiddlyness comes from each request having their own channel. Without being a 100% master of the below code, is the following simpler structure viable?

* LazyBinaryReader (_LBR_) maintains one chan `r.C`
* When someone performs a lookup on the LBR,
   * if there isn't an outstanding request already, we kick one off and create `r.C`. Return `r.C` to the caller.
   * otherwise, there's either a load in progress or the load finished already. return `r.C` to the caller.
* when LBR finishes loading a reader, it stores the result in an atomic `struct{Reader, error}` type thing and closes `r.C`.
* The callers (`getOrLoadReader`, I think) can use `select` to either wait on `r.C` or on their context to die. if `r.C` is signaled, they call `lbr.Result()` to obtain the (reader, error) pair.

`inUse` goes back to being a member of the LBR struct. Unloading waits on `inUse`, then unloads the file. Something like that.

Is that viable, and would a control loop even be necessary?

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:307]: But doesn't it sound funny?

# `cha-chan!`
image

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:377]: > Granted, it will be slower to obtain 100 readers because you need to do 200 channel send+receive serially compared to the 100 concurrent RLock+RUnlocks before.

I think we're talking nanoseconds (maybe a microsecond...) here., it shouldn't be an issue.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:409]: How about we repace the `sync.WaitGroup` with this:
```go
type ctxWaitGroup struct {
	sync.WaitGroup
	once sync.Once
	done chan struct{}
}

func (w *ctxWaitGroup) Wait(ctx context.Context) error {
	w.once.Do(func() {
		go func() {
			w.WaitGroup.Wait()
			close(w.done)
		}()
	})
	select {
	case <-w.done:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}
```

And here do:
```suggestion
			if err := loaded.inUse.Wait(ctx); err != nil {
				unloadPromise.response <- err
			}
```

And we pass a context in the unload request ensuring that there's always a finite timeout (like 10 minutes, etc.)

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:307]: Resolving the comment as this is opinionated and not important.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:172]: Oh I see, how about having a `readerChannel atomic.Value` that holds the `chan readerRequest` and a `unloaded chan struct{}`

Then, when unloading we would do:
```
readerChannel.Set(nil) // or whatever the syntax is
close(done)
```

And `getReader` would look like:

```
func (r *LazyBinaryReader) getReader(ctx context.Context) loadedReader {
	readerReq := readerRequest{response: make(chan loadedReader)}
	select {
	case <-r.unloaded:
		return loadedReader{err: errUnloaded}
	case r.loadedReaderC <- readerReq:
		select {
		...
```

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:157]: I _think_ this goes against the proverb:
> _[Don't communicate by sharing memory, share memory by communicating.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=2m48s)_

I think that the atomic value and signalling is more error prone and harder to follow than a channels approach, where you ask for something, and you get it back.

File: pkg/storegateway/indexheader/reader_pool.go
Comment (narqo) [inline pkg/storegateway/indexheader/reader_pool.go:303]: 👍 I've bumped into [a race-condition][1], which this one fixes

[1]: https://github.com/grafana/mimir/actions/runs/9563917980/job/26363513562

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:157]: I somewhat agree with Oleg. But also would this work with unloading a loaded block - isn't there a race between taking the address of the closed channel and replacing it with a new channel?

I guess we can protect the channel with a mutex, but then we're using 3 sync primitives (is that what you call them?) to achieve the job (atomic, channel, mutex), which looks a bit convoluted.

File: pkg/storegateway/indexheader/reader_pool.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/reader_pool.go:303]: we agreed that it's ok to pull this out of this PR and I can later rebase and resolve conflicts

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:172]: how do you know to start the control loop again after it's been unloaded?

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (56quarters) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:409]: That's a nice solution but it combines a few different types of concurrency primitives and I'd like to see if we can simplify it a bit. It seems weird that there's nothing that does this (waitgroup + context) out of the box.  I don't think we need to block this PR for that so either way is fine to me: log or make a `WaitGroup` wrapper that respects contexts.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:172]: Oh, sure, we'd have to this only when `Close()` is called, which is I assume when we're losing the reference here, is that right?

As I mentioned above:

> Can we just pass a param in the unload request to return from the controlLoop, and pass it in the Close() call?

Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:172]: unfortunately no, from the docs of LazyBinaryReader: https://github.com/grafana/mimir/blob/268fb3cf7fe5bea002c7b0c344b0b5aa8780ac54/pkg/storegateway/indexheader/lazy_binary_reader.go#L177-L178

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go:387]: i think it makes sense. The risk is that we try to load a bad index header which consumes a lot of resources and eventually fails. I can't think of such a realistic case, so I did the change to retry loading it

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go]: so because this happens out of band there isn't a request context (well it's the process startup context, but that's not of a huge use). So I just hardcoded a 30 seconds wait for this. It will be logged here if it times out https://github.com/grafana/mimir/blob/63b5370dbea158847a0c90fa70f1d1ae24d53c57/pkg/storegateway/indexheader/reader_pool.go#L257


Does this sound ok?

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go]: realistically it takes between 0.5 and 23 microseconds to acquire a reader. A single-threaded access is the slowest - 67ns -> 490ns. Generally, the higher the concurrency, the lower the difference between the two approaches

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway/indexheader
                                                  │    before    │                after                 │
                                                  │    sec/op    │    sec/op     vs base                │
NewLazyBinaryReaderConcurrency/concurrency=1-10     67.23n ± 15%   490.40n ± 2%  +629.38% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=2-10     238.7n ±  2%    903.4n ± 2%  +278.55% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=10-10    1.490µ ±  2%    4.527µ ± 1%  +203.83% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=20-10    2.960µ ± 22%    9.269µ ± 0%  +213.18% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=50-10    8.880µ ± 19%   23.920µ ± 1%  +169.38% (p=0.002 n=6)
NewLazyBinaryReaderConcurrency/concurrency=100-10   18.00µ ±  4%    47.51µ ± 1%  +164.01% (p=0.002 n=6)
geomean                                             1.498µ          5.258µ       +250.95%
```

I also ran a modified version of `BenchmarkLabelNames` - one which uses a lazy reader instead of a plain reader. Because that method is so fast we end up with a significant slowdown when using the changes on this branch


```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway/indexheader
                                 │ before-labelnames-with-lazy-reader │   after-labelnames-with-lazy-reader   │
                                 │               sec/op               │    sec/op      vs base                │
LabelNames/20Names100Values-10                           1.554µ ±  2%    2.763µ ±  1%   +77.82% (p=0.002 n=6)
LabelNames/20Names500Values-10                           1.561µ ± 17%    3.334µ ±  4%  +113.58% (p=0.002 n=6)
LabelNames/20Names1000Values-10                          1.551µ ±  1%    3.388µ ±  2%  +118.44% (p=0.002 n=6)
LabelNames/50Names100Values-10                           3.954µ ± 14%    7.139µ ±  2%   +80.55% (p=0.002 n=6)
LabelNames/50Names500Values-10                           3.927µ ±  4%    7.054µ ± 12%   +79.64% (p=0.002 n=6)
LabelNames/50Names1000Values-10                          3.875µ ±  3%    5.289µ ± 16%   +36.49% (p=0.002 n=6)
LabelNames/100Names100Values-10                          8.496µ ±  1%   10.402µ ±  1%   +22.43% (p=0.002 n=6)
LabelNames/100Names500Values-10                          8.493µ ±  1%   10.171µ ±  2%   +19.74% (p=0.002 n=6)
LabelNames/100Names1000Values-10                         8.507µ ±  1%   10.015µ ± 12%   +17.73% (p=0.002 n=6)
LabelNames/200Names100Values-10                          18.11µ ±  0%    20.57µ ±  5%   +13.56% (p=0.002 n=6)
LabelNames/200Names500Values-10                          18.09µ ±  1%    20.50µ ±  1%   +13.29% (p=0.002 n=6)
LabelNames/200Names1000Values-10                         18.15µ ±  1%    20.44µ ±  1%   +12.61% (p=0.002 n=6)
geomean                                                  5.535µ          8.066µ         +45.74%
```

Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go]: the stats are similar for `BenchmarkLabelValuesOffsetsIndexV2` with a lazy reader - 9% slowdown.

Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go]: i couldn't come up with something that's not terribly overcomplicated. Happy to jump on a call with one of your if you're curious what I tried.

I think we should be able to stomach this slow-down. It should only show up for already fast methods. This is the higher-level `BenchmarkBucketStoreLabelValues` which tries to simulate a full store-gateway. The results are mixes

<details><summary>Details</summary>
<p>

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway
                                                                                                                                         │ before-highlevel-benchmarks │      after-highlevel-benchmarks      │
                                                                                                                                         │           sec/op            │    sec/op     vs base                │
BucketStoreLabelValues/no_cache/10-series-matched-with-10-label-values-10                                                                                 658.0µ ±  1%   753.4µ ±  4%  +14.50% (p=0.000 n=12)
BucketStoreLabelValues/no_cache/1000-series-matched-with-1000-label-values-10                                                                             16.27m ±  3%   20.92m ± 14%  +28.57% (p=0.000 n=12)
BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-10-label-values-10                                                                          29.86m ±  2%   29.66m ±  6%        ~ (p=0.799 n=12)
BucketStoreLabelValues/no_cache/1_000_000-series-matched-with-1-label-values-10                                                                           217.9µ ±  7%   239.6µ ±  1%   +9.98% (p=0.000 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/10-series-matched-with-10-label-values-10                                              118.5µ ±  4%   124.3µ ±  6%   +4.87% (p=0.007 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1000-series-matched-with-1000-label-values-10                                          1.127m ±  2%   1.181m ± 12%   +4.79% (p=0.002 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-10-label-values-10                                       29.52m ±  1%   30.42m ±  9%   +3.07% (p=0.000 n=12)
BucketStoreLabelValues/inmemory_cache_(without_label_values_cache)/1_000_000-series-matched-with-1-label-values-10                                        94.45µ ±  1%   99.96µ ±  2%   +5.83% (p=0.000 n=12)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1of1000000-10                                     129.6m ±  7%   126.4m ±  2%   -2.46% (p=0.015 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/10of1000000-10                                    127.0m ±  1%   125.7m ±  1%   -1.00% (p=0.009 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1000000of1000000-10                                1.473 ±  3%    1.493 ±  6%        ~ (p=0.240 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1of1000000-10                           209.2m ± 27%   202.7m ±  4%        ~ (p=0.065 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/10of1000000-10                          234.3m ± 41%   199.5m ±  4%  -14.84% (p=0.004 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1000000of1000000-10                      1.841 ± 47%    1.520 ± 25%  -17.44% (p=0.041 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1of1000000-10                                      151.5m ±  9%   127.9m ±  1%  -15.61% (p=0.002 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/10of1000000-10                                     146.5m ± 14%   127.0m ±  2%  -13.33% (p=0.002 n=6)
BucketStore_Series/1000000SeriesWith1Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1000000of1000000-10                                 1.530 ± 37%    1.512 ±  5%        ~ (p=0.589 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1of10000000-10                                   14.03m ±  3%   13.82m ±  1%   -1.50% (p=0.015 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/100of10000000-10                                 14.05m ± 22%   13.78m ±  2%   -1.95% (p=0.009 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/10000000of10000000-10                            176.7m ± 17%   174.2m ± 11%        ~ (p=0.699 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                         19.50m ±  1%   19.08m ±  6%        ~ (p=0.093 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                       19.19m ±  2%   19.31m ± 17%        ~ (p=0.699 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                  170.4m ±  2%   173.0m ±  3%        ~ (p=0.093 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                                    13.74m ±  1%   14.33m ±  3%   +4.26% (p=0.002 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                                  13.68m ±  4%   14.92m ± 15%   +9.05% (p=0.002 n=6)
BucketStore_Series/100000SeriesWith100Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                             173.7m ± 27%   168.7m ±  6%        ~ (p=1.000 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/1of10000000-10                                   2.174m ± 10%   2.321m ±  4%   +6.75% (p=0.015 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/100of10000000-10                                 2.279m ± 14%   2.379m ± 25%        ~ (p=0.240 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(10K_per_batch)/streamingBatchSize=0/10000000of10000000-10                            89.41m ±  3%   88.06m ± 15%        ~ (p=0.485 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                         2.267m ± 11%   2.194m ±  4%        ~ (p=0.240 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                       2.224m ±  6%   2.216m ±  3%        ~ (p=0.937 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_and_caches_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                  86.48m ±  3%   86.55m ±  2%        ~ (p=0.937 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/1of10000000-10                                    2.499m ±  6%   2.332m ± 34%        ~ (p=0.394 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/100of10000000-10                                  2.473m ±  6%   2.353m ±  7%        ~ (p=0.240 n=6)
BucketStore_Series/1SeriesWith10000000Samples/with_series_streaming_(1K_per_batch)/streamingBatchSize=0/10000000of10000000-10                             88.41m ±  7%   87.52m ±  2%        ~ (p=0.485 n=6)
geomean                                                                                                                                                   20.85m         20.79m         -0.27%

```

</p>
</details>

**Comment by dimitarvdimitrov** [comment]
thanks for the reviews @seizethedave @colega @56quarters. Apologies for dragging this on for so long. I should've resolved most of the suggestions.

At this point the only open point is whether the [perf regression](https://github.com/grafana/mimir/pull/8380#discussion_r1651375692) introduced by channel is worth solving. Can you take a look at the benchmark results in the thread above and let me know what you think?

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [reply pkg/storegateway/indexheader/lazy_binary_reader.go]: I think we can ignore this difference.

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (colega) [inline pkg/storegateway/indexheader/lazy_binary_reader.go:363]: I think 30s is too aggressive here: we've seen slower index lookups.

**Review by colega — APPROVED** [review]

File: pkg/storegateway/indexheader/lazy_binary_reader.go
Comment (dimitarvdimitrov) [reply pkg/storegateway/indexheader/lazy_binary_reader.go]: i realized that timeout is wrong :sob:

if we give up early there will be a goroutine which is waiting on the waitGroup.Wait. If after that comes a request which needs a reader, we will call WaitGroup.Add(1). It is illegal to call Add while Wait is still in progress and WaitGroup will panic.

**Review by 56quarters — APPROVED** [review]
LGTM, nice work!

**Pull request merged**
